File: matrix/c/sol.c
#include<stdio.h>
#include<stdbool.h>
#include<string.h>
#include<stdlib.h>

#define MATRIX_ROWS 3
#define MATRIX_COLS 5

// utility functions
bool checkHorizontal(char matrix[MATRIX_ROWS][MATRIX_COLS], char* target, int row) {
    char *right;
    char *left;
    sprintf(right, "%s%s", matrix[row], matrix[row]);
    sprintf(left, "%s%s", strrev(matrix[row]), strev(matrix[row]));
    return strstr(right, target) || strstr(left, target);
}

bool checkVertical(char matrix[MATRIX_ROWS][MATRIX_COLS], char* target, int col) {
    char *up;
    char *down;
    char *col_s = "";
    getColumn(col_s, matrix, col);
    sprintf(down, "%s%s", col_s, col_s);
    sprintf(up, "%s%s", strrev(col_s), strev(col_s));
    return strstr(up, target) || strstr(down, target);
}

void getColumn(char* s, char matrix[MATRIX_ROWS][MATRIX_COLS], int col) {
    for (int i = 0; i < MATRIX_ROWS; i++) {
        strcat(s, matrix[i][col + 1]);
    }
}


int main(int argc, char const *argv[])
{
    char *target_word = "alpha";
    char matrix[MATRIX_ROWS][MATRIX_COLS] = {"haalp", "bitcl", "ubsup"};

    // solution

    for (int i = 0; i < MATRIX_ROWS; i++) {
        if(checkHorizontal(matrix, target_word, i)) {
            printf("row %d\n", i + 1);
            exit(0);
        }
    }

    for (int i = 0; i < MATRIX_COLS; i++) {
        if(checkVertical(matrix, target_word, i)) {
            printf("col %d\n", i + 1);
            exit(0);
        }
    }
    printf("none\n");
    return 0;
}

File: matrix/sol.pas
uses StrUtils, sysutils;

function checkHorizontal(matrix : Array of String; target_word : String; row : integer):boolean;
var left : String;
    right : String;
    reversed : String;
begin
    right := matrix[row] + matrix[row];
    reversed := ReverseString(matrix[row]);
    left := reversed + reversed;

    checkHorizontal := (Pos(target_word, right) <> 0) Or (Pos(target_word, left) <> 0);
end;

function getColumn(matrix : Array of String; col : integer):String;
    var col_str : String;
        row : String;
begin
    col_str := '';
    for row in matrix do
        col_str := col_str + row[col+1];
    getColumn := col_str
end;

function checkVertical(matrix : Array of String; target_word : String; col : integer):boolean;
var up : String;
    bottom : String;
    reversed : String;
    col_str : String;
begin
    col_str := getColumn(matrix, col);
    bottom := col_str + col_str;
    reversed := ReverseString(col_str);
    up := reversed + reversed;

    checkVertical := (Pos(target_word, bottom) <> 0) Or (Pos(target_word, up) <> 0);
end;
function sol(rows : integer; cols : integer; target_word : String; matrix : Array of String):String;
var i : integer;
begin
    for i:= 0 to rows - 1 do
        if checkHorizontal(matrix, target_word, i) then
            Exit('row ' + IntToStr(i + 1));
    
    for i:= 0 to cols - 1 do
        if checkVertical(matrix, target_word, i) then
            Exit('col ' +IntToStr(i + 1));
    Exit('none');
end;

var
    target_word : String;
    rows : integer;
    cols : integer;
    matrix : Array of String;
    i : integer;
begin
    // read capacity, weights and uses
    read(target_word);
    read(rows);
    read(cols);
    setLength(matrix, rows);

    readln;
    for i:= 0 to rows - 1 do
        readln(matrix[i]);

    writeln(sol(rows, cols, target_word, matrix));
end.